File: Cpp/Lab2_1-2_2/Lab2_2/Ellipse.cpp
#include <stdexcept>
#include <cmath>
#include "Ellipse.h"

double Ellipse::getDistanceToFoci() {
    return isVertical?
        std::sqrt(minorAxis*minorAxis - majorAxis*majorAxis)
        : std::sqrt(majorAxis*majorAxis - minorAxis*minorAxis);
}

Ellipse::Ellipse(double x, double y, double co_vertexDistance, double vertexDistance)
        : Circle(x, y, co_vertexDistance), majorAxis(std::abs(vertexDistance)), isVertical(std::abs(vertexDistance) < std::abs(co_vertexDistance)) {
    if(vertexDistance == 0.0) throw std::invalid_argument("Еліпс не може бути відрізком");
    double dtf = this->getDistanceToFoci();
    this->focus1x = isVertical? x : dtf+x;
    this->focus1y = isVertical? dtf+y : y;
    this->focus2x = isVertical? x : -dtf+x;
    this->focus2y = isVertical? -dtf+y : y;
    this->eccentricity = isVertical? (dtf/co_vertexDistance) : (dtf/vertexDistance);
}

double Ellipse::area() {
    return M_PI * this->majorAxis * this->minorAxis;
}

File: CSharp/Lab2_1-2_2/Lab2_2/Ellipse.cs
namespace Lab2_2;

public class Ellipse : Circle
{
    public double MajorAxis { protected set; get; }
    public (double x, double y) Focus1 { protected set; get; }
    public (double x, double y) Focus2 { protected set; get; }
    public double Eccentricity { protected set; get; }
    public bool IsVertical { protected set; get; }

    public Ellipse(double x, double y, double co_vertexDistance, double vertexDistance)
        : base(x, y, co_vertexDistance)
    {
        this.MajorAxis = Math.Abs(vertexDistance);
        this.IsVertical = Math.Abs(vertexDistance) < Math.Abs(co_vertexDistance);
        if (vertexDistance == 0.0) throw new ArgumentException("Еліпс не може бути відрізком");
        double dtf = GetDistanceToFoci();
        this.Focus1 = IsVertical? (x, dtf+y) : (dtf+x, y);
        this.Focus2 = IsVertical? (x, -dtf+y) : (-dtf+x, y);
        this.Eccentricity = IsVertical ? (dtf / co_vertexDistance) : (dtf / vertexDistance);
    }

    public double Area
    {
        get => Math.PI * this.MajorAxis * this.MinorAxis;
    }

    public double GetDistanceToFoci()
    {
        return IsVertical
            ? Math.Sqrt(MinorAxis * MinorAxis - MajorAxis * MajorAxis)
            : Math.Sqrt(MajorAxis * MajorAxis - MinorAxis * MinorAxis);
    }
}